### PR TITLE
feat: ping metadata

### LINKED ISSUE #445: Implement `ping/1` and `metadata/2` incoming requests.
They are implemented here currently, but with hardcoded metadata: https://github.com/lambdaclass/lambda_ethereum_consensus/blob/edf4c11dea7a2251ceea4d44ee1040c7b5518ede/lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex#L78-L101

Here are the corresponding specs:
- https://github.com/lambdaclass/lambda_ethereum_consensus/blob/main/docs/specs/p2p-interface.md#ping
- https://github.com/lambdaclass/lambda_ethereum_consensus/blob/main/docs/specs/p2p-interface.md#getmetadata-v2

### PR DESCRIPTION
**Pull request by karasakalmt**
Closes #445

### CODE REVIEW

File: lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex [hunk 1]
```diff
@@ -85,9 +85,9 @@ defmodule LambdaEthereumConsensus.P2P.IncomingRequests.Handler do
 
   defp handle_req("metadata/2/ssz_snappy", message_id, _message) do
     # Values are hardcoded
-    with {:ok, payload} <-
-           <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>
-           |> Snappy.compress() do
+    with metadata <- P2P.Metadata.get_metadata(),
```
Comment (Arkenan) [inline lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex:88]: We can leave the metadata assignment outside of the with, as we don't need to match its contents. We can do either:

```elixir
metadata = P2P.Metadata.get_metadata()
with {:ok, payload} <-  Ssz.to_ssz(metadata),
#...
```

or 

```elixir
with {:ok, payload} <- P2P.Metadata.get_metadata() |> Ssz.to_ssz(),
#...

File: lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex [hunk 2]
```diff
@@ -85,9 +85,9 @@ defmodule LambdaEthereumConsensus.P2P.IncomingRequests.Handler do
...
+         {:ok, payload} <- Ssz.to_ssz(metadata),
+         {:ok, payload} <- Snappy.compress(payload) do
```
Comment (Arkenan) [inline lib/lambda_ethereum_consensus/p2p/incoming_requests/handler.ex:90]: Let's not repeat variable names, it's very error prone. We can have "serialized" for the first one and "payload" for the second one.

File: lib/lambda_ethereum_consensus/p2p/metadata.ex [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  @spec get_seq_number() :: SszTypes.uint64()
+  def get_seq_number do
+    [seq_number] = get_metadata_attrs([:seq_number])
+    seq_number
+  end
+
+  @spec get_metadata() :: Metadata.t()
+  def get_metadata do
+    GenServer.call(__MODULE__, :get_metadata, @default_timeout)
```
Comment (Arkenan) [inline lib/lambda_ethereum_consensus/p2p/metadata.ex:24]: Let's not add a timeout. This should always return.

File: lib/lambda_ethereum_consensus/p2p/metadata.ex [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  @impl GenServer
+  def init(metadata) do
+    {:ok, metadata}
+  end
+
+  @impl GenServer
+  def handle_call({:get_metadata_attrs, attrs}, _from, metadata) do
+    values = Enum.map(attrs, &Map.fetch!(metadata, &1))
+    {:reply, values, metadata}
+  end
```
Comment (Arkenan) [inline lib/lambda_ethereum_consensus/p2p/metadata.ex:40]: Are we planning on using this call for something specific? If not, we may want to delete it.

File: lib/lambda_ethereum_consensus/p2p/metadata.ex [hunk 5]
```diff
@@ -0,0 +1,55 @@
+defmodule LambdaEthereumConsensus.P2P.Metadata do
```
Comment (Arkenan) [inline lib/lambda_ethereum_consensus/p2p/metadata.ex:1]: Where is this genserver being started? I think that we should add it to `application.ex`. We also need to add either a `start_link` or `child_spec` public api function for that.

**Review by Arkenan — CHANGES_REQUESTED** [review]

File: lib/lambda_ethereum_consensus/p2p/metadata.ex (hunk 4 above)
Comment (karasakalmt) [reply lib/lambda_ethereum_consensus/p2p/metadata.ex:40]: It is for the ping in first glance but we may need to use it later not sure I can just make a `get_seq_number` instead as well

File: lib/lambda_ethereum_consensus/application.ex [hunk 6]
```diff
@@ -34,6 +34,8 @@ defmodule LambdaEthereumConsensus.Application do
       {LambdaEthereumConsensus.Beacon.PendingBlocks, []},
       {LambdaEthereumConsensus.Beacon.SyncBlocks, []},
       {LambdaEthereumConsensus.P2P.GossipSub, []},
+      {LambdaEthereumConsensus.P2P.Metadata, []},
+      # Start the Endpoint (http/https)
```
Comment (mpaulucci) [inline lib/lambda_ethereum_consensus/application.ex:38]: no need to add this comment

File: lib/lambda_ethereum_consensus/p2p/metadata.ex [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  @moduledoc """
```
Comment (mpaulucci) [inline lib/lambda_ethereum_consensus/p2p/metadata.ex:2]: who sets the data that is stored in this genserver?

Comment (karasakalmt) [reply lib/lambda_ethereum_consensus/p2p/metadata.ex:2]: I don't really know how can i handle that part actually forgot about it. what should i do about here

Comment (mpaulucci) [reply lib/lambda_ethereum_consensus/p2p/metadata.ex:2]: I don't know. Maybe you should take a look at other clients to see where they take this data from.

File: lib/constants.ex [hunk 8]
```diff
@@ -117,4 +117,10 @@ defmodule Constants do
 
   @spec proposer_score_boost() :: non_neg_integer()
   def proposer_score_boost, do: 3
+
+  @spec attestation_subnet_count() :: non_neg_integer()
+  def attestation_subnet_count, do: 2 ** 6
```
Comment (mpaulucci) [inline lib/constants.ex:122]: is this the same constant for every network?

Comment (karasakalmt) [reply lib/constants.ex:122]: took it from
https://github.com/lambdaclass/lambda_ethereum_consensus/blob/main/docs/specs/p2p-interface.md#configuration

Comment (MegaRedHand) [reply lib/constants.ex:122]: Configuration -> changes depending on network

File: lib/lambda_ethereum_consensus/application.ex [hunk 9]
```diff
@@ -29,6 +29,12 @@ defmodule LambdaEthereumConsensus.Application do
       {LambdaEthereumConsensus.Libp2pPort, libp2p_opts},
       {LambdaEthereumConsensus.Store.Db, []},
       {LambdaEthereumConsensus.P2P.Peerbook, []},
+      {LambdaEthereumConsensus.ForkChoice, [checkpoint_sync]},
+      {LambdaEthereumConsensus.P2P.IncomingRequests, []},
+      {LambdaEthereumConsensus.Beacon.PendingBlocks, []},
+      {LambdaEthereumConsensus.Beacon.SyncBlocks, []},
+      {LambdaEthereumConsensus.P2P.GossipSub, []},
+      {LambdaEthereumConsensus.P2P.Metadata, []},
```
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/application.ex:37]: I think there are some merge conflicts here 🤔. Recently we moved some of these to `LambdaEthereumConsensus.Beacon.BeaconNode`

Comment (karasakalmt) [reply lib/lambda_ethereum_consensus/application.ex]: hey I don't know how that happened actually just fixed it

Comment (MegaRedHand) [reply lib/lambda_ethereum_consensus/application.ex]: It's just something `git` does sometimes 😞

File: config/networks/mainnet/config.yaml [hunk 10]
```diff
@@ -96,5 +88,29 @@ DEPOSIT_CONTRACT_ADDRESS: 0x00000000219ab540356cBB839Cbe05303d7705Fa
 
 # Networking
 # ---------------------------------------------------------------
+# `10 * 2**20` (= 10485760, 10 MiB)
+GOSSIP_MAX_SIZE: 10485760
```
Comment (mpaulucci) [inline config/networks/mainnet/config.yaml:92]: where did you get this stuff from?

Comment (karasakalmt) [reply config/networks/mainnet/config.yaml:92]: https://github.com/ethereum/consensus-specs/blob/dev/configs/mainnet.yaml
https://github.com/ethereum/consensus-specs/blob/dev/configs/minimal.yaml

File: lib/lambda_ethereum_consensus/p2p/metadata.ex [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  end
+
+  ##########################
+  ### Private Functions
+  ##########################
+
+  @spec get_metadata_attrs([atom()]) :: [any()]
+  defp get_metadata_attrs(attrs) do
+    GenServer.call(__MODULE__, {:get_metadata_attrs, attrs})
+  end
```
Comment (Arkenan) [inline lib/lambda_ethereum_consensus/p2p/metadata.ex:99]: Noo need to have this as a separate function, let's have the call inlined in the api call.

File: lib/lambda_ethereum_consensus/p2p/metadata.ex [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+  @impl true
+  def handle_call(:get_metadata, _from, metadata) do
+    {:reply, metadata}
+  end
+
+  @impl true
+  def handle_cast({:set_attestation_subnet, i, set}, metadata) do
+    attnets =
+      if set, do: BitVector.set(metadata.attnets, i), else: BitVector.clear(metadata.attnets, i)
```
Comment (Arkenan) [inline lib/lambda_ethereum_consensus/p2p/metadata.ex:69]: Let's have this private function 

```elixir
def set_or_clear(bitvector, i, set)
```

and call it both in hear and in the other cast like follows.

```elixir
attnets = set_or_clear(metadata.attnets, i, set)
```

**Review by Arkenan — DISMISSED** [review]
LGTM, I added a couple of style comments.

**Review by mpaulucci — DISMISSED** [review]

**Comment by karasakalmt** [comment]
@mpaulucci I tried some options to correctly run the CI / Start and stop the node (pull_request) but it doesn't work seems like you have approved with its failure is existing is this something which doesn't need to be fixed?

**Comment by mpaulucci** [comment]
@karasakalmt that should work. Have you tried running the node using `make checkpoint-sync`?

**Review by Arkenan — APPROVED** [review]
LGTM

**Pull request merged**